File: python/math-and-geometry/detect-squares.py
class DetectSquares:

    def __init__(self):
        self.points = defaultdict(lambda: 0)

    def add(self, point: List[int]) -> None:
        self.points[tuple(point)] += 1

    def count(self, point: List[int]) -> int:
        answer = 0
        for p in self.points.keys():
            if (p[0], point[1]) in self.points and (point[0], p[1]) in self.points and p != (p[0], point[1]) and p != (
            point[0], p[1]) and abs(p[0] - point[0]) == abs(p[1] - point[1]):
                answer += self.points[p] * self.points[(p[0], point[1])] * self.points[(point[0], p[1])]

        return answer

# Your DetectSquares object will be instantiated and called as such:
# obj = DetectSquares()
# obj.add(point)
# param_2 = obj.count(point)

File: c++/math-and-geometry/detect-squares.cc
class DetectSquares {
public:
    DetectSquares() {
        
    }
    
    void add(vector<int> point) {
        points_[point[0]][point[1]] += 1;
    }
    
    int count(vector<int> point) {
        int answer = 0;
        
        for (auto [x, map] : points_) {
            for (auto [y, count] : map) {
                if ((points_.contains(x) && points_[x].contains(point[1])) && (points_.contains(point[0]) && points_[point[0]].contains(y)) && (x != point[0] && y != point[1]) && (std::abs(x - point[0]) == std::abs(y - point[1]))) {
                    answer += points_[x][y] * points_[x][point[1]] * points_[point[0]][y];
                } 
            }
        }

        return answer;
    }

private:
    std::unordered_map<int, std::unordered_map<int, int>> points_;
};

/**
 * Your DetectSquares object will be instantiated and called as such:
 * DetectSquares* obj = new DetectSquares();
 * obj->add(point);
 * int param_2 = obj->count(point);
 */
 